pkg/tuf/sign.go

```diff
@@ -1,97 +1,192 @@
+			return fmt.Errorf("cannot add public key target: %v", err)
+		}
+	}
+
+	rootLayoutMeta, err := data.NewFileMeta(bytes.NewBuffer(rootLayout.Custom.InToto.Data), data.NotaryDefaultHashes...)
+	if err != nil {
+		return fmt.Errorf("cannot get root layout meta: %v", err)
+	}
+
+	rootLayoutCustomRawMessage, err := rootLayout.Custom.GetRawMessage()
+	if err != nil {
+		return fmt.Errorf("cannot get raw message for root layout custom: %v", err)
+	}
+
+	linkTarget := client.Target{Name: rootLayout.Filename, Hashes: rootLayoutMeta.Hashes, Length: rootLayoutMeta.Length, Custom: &rootLayoutCustomRawMessage}
+	if err = repo.AddTarget(&linkTarget, data.CanonicalTargetsRole); err != nil {
+		return fmt.Errorf("cannot add root layout target: %v", err)
+	}
+
+	return nil
+}
+
+// clearChangelist clears the notary staging changelist
+func clearChangeList(notaryRepo client.Repository) error {
+	cl, err := notaryRepo.GetChangelist()
+	if err != nil {
+		return err
+	}
+	return cl.Clear("")
+}
+
+// reuse root and top-level targets keys
+func reuseKeys(repo client.Repository, rootKey string) error {
+	if _, err := repo.ListTargets(); err != nil {
 		switch err.(type) {
 		case client.ErrRepoNotInitialized, client.ErrRepositoryNotExist:
 			// Reuse root key.
 			rootKeyIDs, err := importRootKey(rootKey, repo, getPassphraseRetriever())
 			if err != nil {
-				return nil, err
+				return err
 			}
 
 			// NOTE: 2nd variadic argument is to indicate that snapshot is managed remotely.
 			// The impact of a timestamp + snapshot key compromise is not terrible:
 			// https://docs.docker.com/notary/service_architecture/#threat-model
 			if err = repo.Initialize(rootKeyIDs, data.CanonicalSnapshotRole); err != nil {
-				return nil, fmt.Errorf("cannot initialize repo: %v", err)
+				return fmt.Errorf("cannot initialize repo: %v", err)
 			}
 
 			// Reuse targets key.
 			if err = reuseTargetsKey(repo); err != nil {
-				return nil, fmt.Errorf("cannot reuse targets keys: %v", err)
+				return fmt.Errorf("cannot reuse %s keys: %v", data.CanonicalTargetsRole, err)
 			}
 
-		default:
-			return nil, fmt.Errorf("cannot list targets: %v", err)
-		}
-	}
-
-	target, err := client.NewTarget(tag, file, custom)
-	if err != nil {
-		return nil, err
-	}
+			// Reuse targets/releases key.
+			releasesPublicKey, err := reuseReleasesKey(repo)
+			if err != nil {
+				return fmt.Errorf("cannot reuse %s keys: %v", releasesRoleName, err)
+			}
 
-	// TODO - Radu M
-	// decide whether to allow actually passing roles as flags
+			// Delegate to targets/releases.
+			err = delegateToReleases(repo, releasesPublicKey)
+			if err != nil {
+				return fmt.Errorf("cannot delegate to %s: %v", releasesRoleName, err)
+			}
 
-	// If roles is empty, we default to adding to targets
-	if err = repo.AddTarget(target, data.NewRoleList([]string{})...); err != nil {
-		return nil, err
+		default:
+			return fmt.Errorf("cannot list targets: %v", err)
```

Comment:
nit not introduced by this PR: I think the error message should be a bit more descriptive here - yes, the error is because the targets cannot be listed, but it's in the context of signing / reusing keys.